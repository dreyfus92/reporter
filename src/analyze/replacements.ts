import * as replacements from 'module-replacements';
import {
  ReportPluginResult,
  CustomManifest,
  CustomReplacement
} from '../types.js';
import type {FileSystem} from '../file-system.js';
import {getPackageJson} from '../file-system-utils.js';
import {readFile} from 'node:fs/promises';
import {resolve} from 'node:path';

/**
 * Generates a standard URL to the docs of a given rule
 * @param {string} name Rule name
 * @return {string}
 */
export function getDocsUrl(name: string): string {
  return `https://github.com/es-tooling/eslint-plugin-depend/blob/main/docs/rules/${name}.md`;
}

/**
 * Generates a URL for the given path on MDN
 * @param {string} path Docs path
 * @return {string}
 */
export function getMdnUrl(path: string): string {
  return `https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/${path}`;
}

async function loadCustomManifests(
  manifestPaths?: string[]
): Promise<CustomReplacement[]> {
  if (!manifestPaths || manifestPaths.length === 0) {
    return [];
  }

  const customReplacements: CustomReplacement[] = [];

  for (const manifestPath of manifestPaths) {
    try {
      const absolutePath = resolve(manifestPath);
      const manifestContent = await readFile(absolutePath, 'utf8');
      const manifest: CustomManifest = JSON.parse(manifestContent);

      if (manifest.replacements && Array.isArray(manifest.replacements)) {
        customReplacements.push(...manifest.replacements);
      }
    } catch (error) {
      console.warn(
        `Warning: Failed to load custom manifest from ${manifestPath}: ${error}`
      );
    }
  }

  return customReplacements;
}

export async function runReplacements(
  fileSystem: FileSystem,
  customManifests?: string[]
): Promise<ReportPluginResult> {
  const result: ReportPluginResult = {
    messages: []
  };

  const packageJson = await getPackageJson(fileSystem);

  if (!packageJson || !packageJson.dependencies) {
    // No dependencies
    return result;
  }

  // Load custom manifests
  const customReplacements = await loadCustomManifests(customManifests);

  for (const name of Object.keys(packageJson.dependencies)) {
    // Check custom replacements first
    let replacement = customReplacements.find(
      (replacement) => replacement.moduleName === name
    );

    // If no custom replacement found, check built-in replacements
    if (!replacement) {
      replacement = replacements.all.moduleReplacements.find(
        (replacement) => replacement.moduleName === name
      );
    }

    if (!replacement) {
      continue;
    }

    if (replacement.type === 'none') {
      result.messages.push({
        severity: 'warning',
        score: 0,
        message: `Module "${name}" can be removed, and native functionality used instead`
      });
    } else if (replacement.type === 'simple') {
      result.messages.push({
        severity: 'warning',
        score: 0,
        message: `Module "${name}" can be replaced. ${replacement.replacement || 'See documentation for details'}.`
      });
    } else if (replacement.type === 'native') {
      const mdnPath = replacement.mdnPath
        ? getMdnUrl(replacement.mdnPath)
        : undefined;
      // TODO (43081j): support `nodeVersion` here, check it against the
      // packageJson.engines field, if there is one.
      const message = `Module "${name}" can be replaced with native functionality. Use "${replacement.replacement || 'native alternative'}" instead.`;
      const fullMessage = mdnPath
        ? `${message} You can read more at ${mdnPath}.`
        : message;
      result.messages.push({
        severity: 'warning',
        score: 0,
        message: fullMessage
      });
    } else if (replacement.type === 'documented') {
      const docUrl = replacement.docPath
        ? getDocsUrl(replacement.docPath)
        : undefined;
      const message = `Module "${name}" can be replaced with a more performant alternative.`;
      const fullMessage = docUrl
        ? `${message} See the list of available alternatives at ${docUrl}.`
        : message;
      result.messages.push({
        severity: 'warning',
        score: 0,
        message: fullMessage
      });
    }
  }

  return result;
}
