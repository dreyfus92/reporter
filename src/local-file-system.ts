import type {FileSystem} from './file-system.js';
import * as path from 'node:path';
import * as fs from 'node:fs/promises';
import {fileSystemLogger} from './logger.js';
import {fdir} from 'fdir';
import {readFile, stat} from 'node:fs/promises';
import {normalizePath} from './utils/path.js';

export class LocalFileSystem implements FileSystem {
  #root: string;
  #logger = fileSystemLogger;

  constructor(root: string) {
    this.#root = root;
  }

  async getRootDir(): Promise<string> {
    return this.#root;
  }

  async listPackageFiles(): Promise<string[]> {
    const nodeModulesPath = path.join(this.#root, 'node_modules');

    try {
      await fs.access(nodeModulesPath);
      const crawler = new fdir()
        .withFullPaths()
        .withSymlinks()
        .filter((filePath) => normalizePath(filePath).endsWith('/package.json'))
        .crawl(nodeModulesPath);
      const files = await crawler.withPromise();
      return files.map((file) => {
        const relativePath = path.relative(this.#root, file);
        return '/' + normalizePath(relativePath);
      });
    } catch {
      return [];
    }
  }

  async readFile(filePath: string): Promise<string> {
    return await readFile(path.join(this.#root, filePath), 'utf8');
  }

  async getInstallSize(): Promise<number> {
    const nodeModulesPath = path.join(this.#root, 'node_modules');

    let installSize = 0;

    try {
      await fs.access(nodeModulesPath);

      // TODO (43081j): we traverse the file system twice. Once here,
      // and once when finding the package.json files. It may be worth
      // caching some things one day so we only traverse once.
      const crawler = new fdir()
        .withFullPaths()
        .withSymlinks()
        .crawl(nodeModulesPath);
      const files = await crawler.withPromise();
      for (const filePath of files) {
        try {
          const stats = await stat(filePath);
          installSize += stats.size;
        } catch {
          this.#logger.debug('Error getting file stats for:', filePath);
        }
      }
    } catch {
      this.#logger.debug('No node_modules directory found');
    }

    return installSize;
  }
}
