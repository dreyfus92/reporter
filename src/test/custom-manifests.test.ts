import {describe, it, expect} from 'vitest';
import {runReplacements} from '../analyze/replacements.js';
import {LocalFileSystem} from '../local-file-system.js';
import {join} from 'node:path';
import {fileURLToPath} from 'node:url';
import {dirname} from 'node:path';

const __dirname = dirname(fileURLToPath(import.meta.url));

describe('Custom Manifests', () => {
  it('should load and use custom manifest files', async () => {
    const testDir = join(__dirname, '../../test/fixtures/basic-chalk');
    const fileSystem = new LocalFileSystem(testDir);
    const customManifestPath = join(
      __dirname,
      '../../test/fixtures/custom-manifest.json'
    );

    const result = await runReplacements(fileSystem, [customManifestPath]);

    // Should have messages from custom manifest
    expect(result.messages.length).toBeGreaterThan(0);

    // Check that custom replacement messages are included
    const hasCustomMessage = result.messages.some(
      (msg) =>
        msg.message.includes('chalk') ||
        msg.message.includes('lodash') ||
        msg.message.includes('moment') ||
        msg.message.includes('request') ||
        msg.message.includes('bluebird')
    );

    expect(hasCustomMessage).toBe(true);
  });

  it('should handle invalid manifest files gracefully', async () => {
    const testDir = join(__dirname, '../../test/fixtures/basic-chalk');
    const fileSystem = new LocalFileSystem(testDir);
    const invalidManifestPath = 'non-existent-file.json';

    const result = await runReplacements(fileSystem, [invalidManifestPath]);

    // Should still work without crashing
    expect(result.messages).toBeDefined();
  });

  it('should prioritize custom replacements over built-in ones', async () => {
    const testDir = join(__dirname, '../../test/fixtures/basic-chalk');
    const fileSystem = new LocalFileSystem(testDir);
    const customManifestPath = join(
      __dirname,
      '../../test/fixtures/custom-manifest.json'
    );

    const resultWithCustom = await runReplacements(fileSystem, [
      customManifestPath
    ]);
    const resultWithoutCustom = await runReplacements(fileSystem);

    // Custom manifest should provide additional or different messages
    expect(resultWithCustom.messages.length).toBeGreaterThanOrEqual(
      resultWithoutCustom.messages.length
    );
  });

  it('should load multiple manifest files', async () => {
    const testDir = join(__dirname, '../../test/fixtures/basic-chalk');
    const fileSystem = new LocalFileSystem(testDir);
    const manifest1Path = join(
      __dirname,
      '../../test/fixtures/custom-manifest.json'
    );
    const manifest2Path = join(
      __dirname,
      '../../test/fixtures/custom-manifest-2.json'
    );

    const result = await runReplacements(fileSystem, [
      manifest1Path,
      manifest2Path
    ]);

    // Should have messages from both manifests
    expect(result.messages.length).toBeGreaterThan(0);

    // Check that replacements from both manifests are included
    const hasChalkMessage = result.messages.some((msg) =>
      msg.message.includes('chalk')
    );
    
    // Note: express won't be found since it's not in the test fixture dependencies
    // but chalk should be found
    expect(hasChalkMessage).toBe(true);
  });
});
