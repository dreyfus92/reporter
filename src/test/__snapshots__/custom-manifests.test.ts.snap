// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Custom Manifests > should handle invalid manifest files gracefully 1`] = `[]`;

exports[`Custom Manifests > should load and use custom manifest files 1`] = `
[
  {
    "message": "Module "@e18e/fake-0" can be replaced. Use picocolors, kleur, or native console styling.",
    "score": 0,
    "severity": "warning",
  },
  {
    "message": "Module "@e18e/fake-1" can be replaced. Use native JavaScript methods or specific lodash functions.",
    "score": 0,
    "severity": "warning",
  },
  {
    "message": "Module "@e18e/fake-2" can be replaced with native functionality. Use "Intl.DateTimeFormat or Date methods" instead. Required Node >= 12.0.0. You can read more at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat.",
    "score": 0,
    "severity": "warning",
  },
  {
    "message": "Module "@e18e/fake-3" can be replaced with a more performant alternative. See the list of available alternatives at https://github.com/es-tooling/eslint-plugin-depend/blob/main/docs/rules/request-alternatives.md.",
    "score": 0,
    "severity": "warning",
  },
  {
    "message": "Module "@e18e/fake-4" can be removed, and native functionality used instead",
    "score": 0,
    "severity": "warning",
  },
]
`;

exports[`Custom Manifests > should load multiple manifest files 1`] = `
[
  {
    "message": "Module "@e18e/fake-0" can be replaced. Use picocolors, kleur, or native console styling.",
    "score": 0,
    "severity": "warning",
  },
  {
    "message": "Module "@e18e/fake-1" can be replaced. Use native JavaScript methods or specific lodash functions.",
    "score": 0,
    "severity": "warning",
  },
  {
    "message": "Module "@e18e/fake-2" can be replaced with native functionality. Use "Intl.DateTimeFormat or Date methods" instead. Required Node >= 12.0.0. You can read more at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat.",
    "score": 0,
    "severity": "warning",
  },
  {
    "message": "Module "@e18e/fake-3" can be replaced with a more performant alternative. See the list of available alternatives at https://github.com/es-tooling/eslint-plugin-depend/blob/main/docs/rules/request-alternatives.md.",
    "score": 0,
    "severity": "warning",
  },
  {
    "message": "Module "@e18e/fake-4" can be removed, and native functionality used instead",
    "score": 0,
    "severity": "warning",
  },
  {
    "message": "Module "@e18e/fake-5" can be replaced. Use Fastify, Koa, or native Node.js http module.",
    "score": 0,
    "severity": "warning",
  },
  {
    "message": "Module "@e18e/fake-6" can be removed, and native functionality used instead",
    "score": 0,
    "severity": "warning",
  },
]
`;

exports[`Custom Manifests > should prioritize custom replacements over built-in ones 1`] = `
{
  "withCustom": [
    {
      "message": "Module "@e18e/fake-0" can be replaced. Use picocolors, kleur, or native console styling.",
      "score": 0,
      "severity": "warning",
    },
    {
      "message": "Module "@e18e/fake-1" can be replaced. Use native JavaScript methods or specific lodash functions.",
      "score": 0,
      "severity": "warning",
    },
    {
      "message": "Module "@e18e/fake-2" can be replaced with native functionality. Use "Intl.DateTimeFormat or Date methods" instead. Required Node >= 12.0.0. You can read more at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat.",
      "score": 0,
      "severity": "warning",
    },
    {
      "message": "Module "@e18e/fake-3" can be replaced with a more performant alternative. See the list of available alternatives at https://github.com/es-tooling/eslint-plugin-depend/blob/main/docs/rules/request-alternatives.md.",
      "score": 0,
      "severity": "warning",
    },
    {
      "message": "Module "@e18e/fake-4" can be removed, and native functionality used instead",
      "score": 0,
      "severity": "warning",
    },
  ],
  "withoutCustom": [],
}
`;
